### ISSUE TITLE
Return extra runtime statistics information immediately

### ISSUE THREAD

**Issue by Chaho12** [trinodb/trino-go-client#12]
Unlike other other drivers which simply uses http request and gets runtime statistics information using HTTP protocol, Presto-Go-Client uses https://golang.org/pkg/database/sql/driver to implement and returns rows only after the query completes

@losipiuk suggests using named query arguments to pass callback to implementation of QueryContext method.
- https://github.com/prestosql/presto-go-client/commit/be6bb15edd764217e6626ceb14f024bdf33cc4d9

### FYI

After setting connection by sql.Open("presto", dsn) whole process of querying data is as follows. These functions are all within go SQL.

`sql.Query -> QueryContext -> query -> queryDC -> rowsiFromStatement -> ctxDriverStmtQuery ->  siCtx.QueryContext(ctx, nvdargs) ->` then returns rows
- queryDC : queryDC executes a query on the given connection.
- siCtx.QueryContext(ctx, nvdargs) : presto-go-client QueryContext function

Relevant slack discussion: https://prestosql.slack.com/archives/CFLB9AMBN/p1597988801033400

**Comment by losipiuk** [comment]
Thanks for the issue. 

I would suggest to model the object representing statistics after what we have for JDBC https://github.com/prestosql/presto/blob/c74b775e103a7a239aafef1b7ca15fe6370495fe/presto-jdbc/src/main/java/io/prestosql/jdbc/QueryStats.java.
And the callback interface could be:

```go
type QueryStatsCallback interface{
   // called on query stats update. Final call will be done just after query execution finished.
   StatsUpdated(QueryStats)

   // value in milliseconds determining how often the stats callback should be called. If 0 it will be called as frequently as we get new stats from the server.
   GetCallBackPeriod() int64
}
```

Then we can register the stats via extra argument passed to `QueryContext`:
```go
db.Query("SELECT 2", sql.Named("X-Query-Stats-Callback", callback))
```

Alternatively we can drop `GetCallBackPeriod` from the interface and use extra `X-Query-Stats-Callback-Period` parameter instead.

@Chaho12 @ebyhr WDYT?

**Comment by Chaho12** [comment]
@losipiuk can you checkout this code? I added `ProgressPercentage` and added your callback code inside `fetch` method. 
- [fetch](https://github.com/prestosql/presto-go-client/blob/master/presto/presto.go#L608) is called inside QueryContext and this is where the actual HTTP Get is called. 

https://github.com/Chaho12/presto-go-client/commit/554b06ea2e67beb467da67b94e4a6e22126ae91a

### test code result

```
GOROOT=/usr/local/go #gosetup
GOPATH=/Users/user/go #gosetup
/usr/local/go/bin/go test -c -o /private/var/folders/5d/mbcxtdy13sj4vcwprc7pphpc0000gn/T/___TestQueryWithCallbackAsParam_in_github_com_prestosql_presto_go_client_presto github.com/prestosql/presto-go-client/presto #gosetup
/usr/local/go/bin/go tool test2json -t /private/var/folders/5d/mbcxtdy13sj4vcwprc7pphpc0000gn/T/___TestQueryWithCallbackAsParam_in_github_com_prestosql_presto_go_client_presto -test.v -test.run ^TestQueryWithCallbackAsParam$ #gosetup
=== RUN   TestQueryWithCallbackAsParam
query started
query progress
[], QUEUED, 0 
query progress
[], QUEUED, 0 
query finished
[[2]], RUNNING, 0 
2
--- PASS: TestQueryWithCallbackAsParam (0.06s)
PASS

Process finished with exit code 0
```

**Comment by losipiuk** [comment]
Thanks.

The approach looks valid, though I would suggest to go with simpler callback with just one method. We can document that method would be called immediately after query starts and just after it ends. And on regular intervals during execution.
Also passing `queryResponse` as callback parameter will not work as it is private structure. I think we should have a specific struct to be used just for callback purpose, and construct it based on `queryResponse`.

**Comment by Chaho12** [comment]
I'll look further into it tomorow :)

**Comment by losipiuk** [comment]
@Chaho12 please let me know if you need help with that.

**Comment by Chaho12** [comment]
@losipiuk Thx for asking. I didn't have any time to work on this since my last comment and won't have time until later in October. I have to work on other project first and this part of project is put on hold.  
If you have time, please feel free continue without me :). I'll come back after am done with urgent matters.

**Comment by electrum** [comment]
JDBC doesn’t bother with a callback interval. It simply calls for every request made to the server. This is typically once per second while the query is executing.

**Comment by Chaho12** [comment]
@electrum When you mean JDBC doesn't bother with a callback interval, do you suggest that callback interval should be set as once per second and should not change by user's request? [As of now](https://github.com/prestosql/presto-go-client/blob/master/presto/presto.go#L397-L433), query makes request to the server without any intervals unless it returns error code. 
This seems to be a burden (am designing so that callback is called once per every request made to server), shouldn't `line 400` `timer := time.NewTimer(0)` should be `timer := time.NewTimer(time.Second * 1)` so that it waits before a request is made?

**Comment by Chaho12** [comment]
@electrum  @losipiuk  
What do you guys think of [line 400](https://github.com/prestosql/presto-go-client/blob/master/presto/presto.go#L400) where `timer := time.NewTimer(0)` should be changed because it makes request to server immediately when the function is called? Maybe add a timer param with default value of 0? 
- This `roundtrip` function is called within a recursive function called `fetch` which calls `roundtrip` function without any  delays until `qr.nextURI` is empty (delay if and only if `http.StatusServiceUnavailable`)
- https://github.com/prestosql/presto-go-client/blob/master/presto/presto.go#L759
